### ISSUE TITLE
Address 0x8 error

### ISSUE THREAD

**Issue by nandofernandesneto** [rstudio/keras#816]
While trying to test the code snipped provided here: https://medium.com/@alexeybnk/can-reinforcement-learning-trade-stock-implementation-in-r-8cd54d13165c 

I have gotten the following error:

> critic_model_v <-
+   critic_input %>%
+   common_layer_dense_1 %>%
+   critic_layer_dense_v_1 %>%
+   critic_layer_dense_v_2 %>%
+   critic_layer_dense_v_3 %>%
+   v_normalize_layer

 *** caught segfault ***
address 0x8, cause 'memory not mapped'

Traceback:
 1: py_call_impl(callable, dots$args, dots$keywords)
 2: layer(object)
 3: compose_layer.python.builtin.object(object, layer)
 4: compose_layer(object, layer)
 5: create_layer(v_CustomLayer, object, list(name = "v_normalize_layer"))
 6: v_normalize_layer(.)
 7: function_list[[k]](value)
 8: withVisible(function_list[[k]](value))
 9: freduce(value, `_function_list`)
10: `_fseq`(`_lhs`)
11: eval(quote(`_fseq`(`_lhs`)), env, env)
12: eval(quote(`_fseq`(`_lhs`)), env, env)
13: withVisible(eval(quote(`_fseq`(`_lhs`)), env, env))
14: critic_input %>% common_layer_dense_1 %>% critic_layer_dense_v_1 %>%     critic_layer_dense_v_2 %>% critic_layer_dense_v_3 %>% v_normalize_layer

Possible actions:
1: abort (with core dump, if enabled)
2: normal R exit
3: exit R without saving workspace
4: exit R saving workspace

Any ideas of what could it be?

Thanks in advance,

Best,

Fernando

**Comment by nandofernandesneto** [comment]
PS: the code works on my windows machine ... but it doesn't on my linux server

**Comment by dfalbel** [comment]
We recently fixed a bug related to a segfault with custom layers. Which version of keras and tensorflow are you using?

**Comment by nandofernandesneto** [comment]
Hi, thanks for replying.

This is my session info ()

R version 3.5.1 (2018-07-02)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 18.04.2 LTS

Matrix products: default
BLAS: /opt/microsoft/ropen/3.5.1/lib64/R/lib/libRblas.so
LAPACK: /opt/microsoft/ropen/3.5.1/lib64/R/lib/libRlapack.so

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C               LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8     LC_MONETARY=en_US.UTF-8   
 [6] LC_MESSAGES=en_US.UTF-8    LC_PAPER=en_US.UTF-8       LC_NAME=C                  LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] SparseM_1.77             keras_2.2.4.1.9001       tfprobability_0.0.0.9000 tensorflow_1.13.1.9000   RevoUtils_11.0.1        

loaded via a namespace (and not attached):
 [1] Rcpp_0.12.18           lattice_0.20-35        zeallot_0.1.0          grid_3.5.1             R6_2.2.2              
 [6] jsonlite_1.5           magrittr_1.5           tfruns_1.3             whisker_0.3-2          Matrix_1.2-14         
[11] reticulate_1.12.0-9001 generics_0.0.2         tools_3.5.1            yaml_2.2.0             compiler_3.5.1        
[16] base64enc_0.1-3

**Comment by nandofernandesneto** [comment]
This is the code, for reproduction purposes (of the error)

`# configure critic NN ------------

library('keras')
library('R6')

learning_rate <- 1e-3
state_names_length <- 12 # just for example
acts <- c("buy", "hold", "sell")

a_CustomLayer <- R6::R6Class(
  "CustomLayer"
  , inherit = KerasLayer
  , public = list(
    
    call = function(x, mask = NULL) {
      x - k_mean(x, axis = 2, keepdims = T)
    }
    
  )
)

a_normalize_layer <- function(object) {
  create_layer(a_CustomLayer, object, list(name = 'a_normalize_layer'))
}

v_CustomLayer <- R6::R6Class(
  "CustomLayer"
  , inherit = KerasLayer
  , public = list(
    
    call = function(x, mask = NULL) {
      k_concatenate(list(x, x, x), axis = 2)
    }
    
    , compute_output_shape = function(input_shape) {
      
      output_shape = input_shape
      output_shape[[2]] <- input_shape[[2]] * 3L
      
      output_shape
    }
  )
)

v_normalize_layer <- function(object) {
  create_layer(v_CustomLayer, object, list(name = 'v_normalize_layer'))
}

noise_CustomLayer <- R6::R6Class(
  "CustomLayer"
  , inherit = KerasLayer
  , lock_objects = FALSE
  , public = list(
    
    initialize = function(output_dim) {
      self$output_dim <- output_dim
    }
    
    , build = function(input_shape) {
      
      self$input_dim <- input_shape[[2]]
      
      sqr_inputs <- self$input_dim ** (1/2)
      
      self$sigma_initializer <- initializer_constant(.5 / sqr_inputs)
      
      self$mu_initializer <- initializer_random_uniform(minval = (-1 / sqr_inputs), maxval = (1 / sqr_inputs))
      
      self$mu_weight <- self$add_weight(
        name = 'mu_weight', 
        shape = list(self$input_dim, self$output_dim),
        initializer = self$mu_initializer,
        trainable = TRUE
      )
      
      self$sigma_weight <- self$add_weight(
        name = 'sigma_weight', 
        shape = list(self$input_dim, self$output_dim),
        initializer = self$sigma_initializer,
        trainable = TRUE
      )
      
      self$mu_bias <- self$add_weight(
        name = 'mu_bias', 
        shape = list(self$output_dim),
        initializer = self$mu_initializer,
        trainable = TRUE
      )
      
      self$sigma_bias <- self$add_weight(
        name = 'sigma_bias', 
        shape = list(self$output_dim),
        initializer = self$sigma_initializer,
        trainable = TRUE
      )
      
    }
    
    , call = function(x, mask = NULL) {
      
      #sample from noise distribution
      
      e_i = k_random_normal(shape = list(self$input_dim, self$output_dim))
      e_j = k_random_normal(shape = list(self$output_dim))
      
      
      #We use the factorized Gaussian noise variant from Section 3 of Fortunato et al.
      
      eW = k_sign(e_i) * (k_sqrt(k_abs(e_i))) * k_sign(e_j) * (k_sqrt(k_abs(e_j)))
      eB = k_sign(e_j) * (k_abs(e_j) ** (1/2))
      
      
      #See section 3 of Fortunato et al.
      
      noise_injected_weights = k_dot(x, self$mu_weight + (self$sigma_weight * eW))
      noise_injected_bias = self$mu_bias + (self$sigma_bias * eB)
      output = k_bias_add(noise_injected_weights, noise_injected_bias)
      
      output
      
    }
    
    , compute_output_shape = function(input_shape) {
      
      output_shape <- input_shape
      output_shape[[2]] <- self$output_dim
      
      output_shape
      
    }
  )
)

noise_add_layer <- function(object, output_dim) {
  create_layer(
    noise_CustomLayer
    , object
    , list(
      name = 'noise_add_layer'
      , output_dim = as.integer(output_dim)
      , trainable = T
    )
  )
}

critic_input <- layer_input(
  shape = c(as.integer(state_names_length))
  , name = 'critic_input'
)

common_layer_dense_1 <- layer_dense(
  units = 20
  , activation = "tanh"
)

critic_layer_dense_v_1 <- layer_dense(
  units = 10
  , activation = "tanh"
)

critic_layer_dense_v_2 <- layer_dense(
  units = 5
  , activation = "tanh"
)

critic_layer_dense_v_3 <- layer_dense(
  units = 1
  , name = 'critic_layer_dense_v_3'
)

critic_layer_dense_a_1 <- layer_dense(
  units = 10
  , activation = "tanh"
)

# critic_layer_dense_a_2 <- layer_dense(
#      units = 5
#      , activation = "tanh"
# )

critic_layer_dense_a_3 <- layer_dense(
  units = length(acts)
  , name = 'critic_layer_dense_a_3'
)

critic_model_v <-
  critic_input %>%
  common_layer_dense_1 %>%
  critic_layer_dense_v_1 %>%
  critic_layer_dense_v_2 %>%
  critic_layer_dense_v_3 %>%
  v_normalize_layer

critic_model_a <-
  critic_input %>%
  common_layer_dense_1 %>%
  critic_layer_dense_a_1 %>%
  #critic_layer_dense_a_2 %>%
  noise_add_layer(output_dim = 5) %>%
  critic_layer_dense_a_3 %>%
  a_normalize_layer

critic_output <-
  layer_add(
    list(
      critic_model_v
      , critic_model_a
    )
    , name = 'critic_output'
  )

critic_model_1  <- keras_model(
  inputs = critic_input
  , outputs = critic_output
)

critic_optimizer = optimizer_adam(lr = learning_rate)

keras::compile(
  critic_model_1
  , optimizer = critic_optimizer
  , loss = 'mse'
  , metrics = 'mse'
)

train.x <- rnorm(state_names_length * 10)

train.x <- array(train.x, dim = c(10, state_names_length))

predict(critic_model_1, train.x)

critic_model_2 <- critic_model_1`

**Comment by dfalbel** [comment]
I can't reproduce the error on both my mac and linux box. Can you try re-installing keras from github? Also, can you please send your `tf_config()`?

**Comment by nandofernandesneto** [comment]
TensorFlow v1.13.1 (~/anaconda3/envs/Tensorflow_for_R/lib/python3.7/site-packages/tensorflow_probability)
Python v3.7 (~/anaconda3/envs/Tensorflow_for_R/bin/python)

**Comment by nandofernandesneto** [comment]
I have just installed the github version. It is working. Thanks!
